pkg/scheduler/sorters.go

```diff
@@ -47,39 +39,101 @@ func sortQueue(queues []*SchedulingQueue, sortType SortType) {
 			return comp < 0
 		})
 	}
+	metrics.GetSchedulerMetrics().ObserveQueueSortingLatency(sortingStart)
 }
 
-func sortApplications(apps []*SchedulingApplication, sortType SortType, globalResource *resources.Resource) {
-	// TODO add latency metric
+func sortApplications(apps map[string]*SchedulingApplication, sortType policies.SortPolicy, globalResource *resources.Resource) []*SchedulingApplication {
+	sortingStart := time.Now()
+	var sortedApps []*SchedulingApplication
 	switch sortType {
-	case FairSortPolicy:
+	case policies.FairSortPolicy:
+		sortedApps = filterOnPendingResources(apps)
 		// Sort by usage
-		sort.SliceStable(apps, func(i, j int) bool {
-			l := apps[i]
-			r := apps[j]
+		sort.SliceStable(sortedApps, func(i, j int) bool {
+			l := sortedApps[i]
+			r := sortedApps[j]
 			return resources.CompUsageRatio(l.getAssumeAllocated(), r.getAssumeAllocated(), globalResource) < 0
 		})
-	case FifoSortPolicy:
+	case policies.FifoSortPolicy:
+		sortedApps = filterOnPendingResources(apps)
 		// Sort by submission time oldest first
-		sort.SliceStable(apps, func(i, j int) bool {
-			l := apps[i]
-			r := apps[j]
+		sort.SliceStable(sortedApps, func(i, j int) bool {
+			l := sortedApps[i]
+			r := sortedApps[j]
 			return l.ApplicationInfo.SubmissionTime < r.ApplicationInfo.SubmissionTime
 		})
+	case policies.StateAwarePolicy:
+		sortedApps = stateAwareFilter(apps)
+		// Sort by submission time oldest first
+		sort.SliceStable(sortedApps, func(i, j int) bool {
+			l := sortedApps[i]
+			r := sortedApps[j]
+			return l.ApplicationInfo.SubmissionTime < r.ApplicationInfo.SubmissionTime
+		})
+	}
+	metrics.GetSchedulerMetrics().ObserveAppSortingLatency(sortingStart)
+	return sortedApps
+}
+
+func filterOnPendingResources(apps map[string]*SchedulingApplication) []*SchedulingApplication {
+	filteredApps := make([]*SchedulingApplication, 0)
+	for _, app := range apps {
+		// Only look at app when pending-res > 0
+		if resources.StrictlyGreaterThanZero(app.GetPendingResource()) {
+			filteredApps = append(filteredApps, app)
+		}
+	}
+	return filteredApps
+}
+
+// This filter only allows one (1) application with a state that is not running in the list of candidates.
+// The preference is a state of Starting. If we can not find an app with a starting state we will use an app
+// with an Accepted state. However if there is an app with a Starting state even with no pending resource
+// requests, no Accepted apps can be scheduled. An app in New state does not have any asks and can never be
+// scheduled.
+func stateAwareFilter(apps map[string]*SchedulingApplication) []*SchedulingApplication {
```

Comment:
one issue for this algorithm is
If I have a bunch of apps, one is `Starting` and the rest is `Running`, then only the `Starting` app can be filtered out. If my `Starting` app happens to be a single pod app. Even the pod is running,  it will need to wait for at least 5 minutes before other apps can be scheduled.

Reply:
Basically, I think the `Running` state should not be depending on the fact that there is a 2nd allocation that happened for the app. Instead, we should read the info from the app, each app can have a field, e.g `org.apache.yunikorn/minTaskNum` (default=1). When this is 1, then once 1 allocation is made, we can say the app is `Running`. Later on, we can leverage this field for FIFO based Gang scheduling.

Reply:
We can not use this sorting for gang scheduling, there is no dependency between the asks. That means that if the app sets the minimum to 5 the scheduler will just wait with progressing the state of the application. Allocations will still be 1 at a time and can be interrupted by other apps.. This means that there is no guarantee that the scheduler will allocate one or more of them at the same time or even within a limited amount of time.

Setting a default of 1 is not the correct setting. The default would mean that the Starting state is always skipped unless the submitter says otherwise. In essence you turn off the Starting state and thus state aware sorting, except for those apps that want to use it. You need to set a default that will force all apps to go through the starting state for 1 allocation _unless_ the app specifies differently.

Reply:
do we have any other way to resolve this problem? this is a blocker to me

Reply:
Which part is a blocker: the gang side is not possible to solve as this sorting was never designed to solve gang scheduling. Gang scheduling is also not related to the state change that is or is not triggered. A gang would be a set of asks to return only if that set is complete. The app could ask once or more times for a gang, it might also not be the first request coming in.

Two things:
I chose the timeout of 5 minutes, no idea if that is too long or not as the default.

For the single container/pod we can leverage a tag or label on the application. It will not change anything on the filtering side. The only impact will be in the app when moving into and out of the `Starting` state. If the tag or label is set the app itself quickly auto progresses using a `RunApplication` event. We do not skip the state we move the app out of the `Starting` state almost immediately. That solves the single container/pod issue.

Requiring or even allowing multiple allocations to move out of `Starting` is not really a use case that will show up as far as I can assess. It does not cover a gang type request, it does also not fit in with any application type I know.

Reply:
We can use a simple label like this: `yunikorn.apache.org/singleAllocation`
Value set I don't think matters, it is passed as a string. The application will progress out of the `Starting` state as follows:
- Any non empty value --> there is just one allocation, _immediate_ auto progress.
- Not present or empty --> progress on the a second allocation or after the default timeout.

Reply:
based on the discussion. Let's group un-named pods into one single app per namespace. this can resolve the issue.